Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -243,219 +242,175 @@ ActsExamples::ProcessCode ActsExamples::VertexPerformanceWriter::writeT(
+  ACTS_INFO("Number of reco track-associated truth particles in event : "
+            << recoParticles.size());
+  ACTS_INFO("Maximum number of reconstructible primary vertices : "
+            << m_nVtxReconstructable);
 
   // We compare the reconstructed momenta to the true momenta at the vertex. For
   // this, we propagate the reconstructed tracks to the PCA of the true vertex
   // position. Setting up propagator:
   Acts::EigenStepper<> stepper(m_cfg.bField);
   using Propagator = Acts::Propagator<Acts::EigenStepper<>>;
   auto propagator = std::make_shared<Propagator>(stepper);
+  // Setting the geometry/magnetic field context for the event
+  Acts::PropagatorOptions pOptions(ctx.geoContext, ctx.magFieldContext);
 
   // Loop over reconstructed vertices and see if they can be matched to a true
   // vertex.
   for (const auto& vtx : vertices) {
     // Reconstructed tracks that contribute to the reconstructed vertex
-    const auto tracksAtVtx = vtx.tracks();
+    const auto& tracksAtVtx = vtx.tracks();
+    // Input tracks matched to `tracksAtVtx`
+    std::vector<std::uint32_t> trackIndices;
 
     // Containers for storing truth particles and truth vertices that contribute
     // to the reconstructed vertex
     SimParticleContainer particleAtVtx;
-    std::vector<SimBarcode> contributingTruthVertices;
+    SimBarcodeContainer contributingTruthVertices;
 
     if (m_cfg.useTracks) {
       for (const auto& trk : tracksAtVtx) {
         // Track parameters before the vertex fit
         const Acts::BoundTrackParameters& origTrack =
             *trk.originalParams.as<Acts::BoundTrackParameters>();
 
+        bool foundMatchingParticle = false;
+
         // Finding the matching parameters in the container of all track
-        // parameters. This allows us to identify the corresponding particle,
-        // since we expect trackParameters and associatedTruthParticles to
-        // align.
-        bool foundMatchingParams = false;
-        for (std::size_t i = 0; i < trackParameters.size(); ++i) {
-          const auto& params = trackParameters[i].parameters();
-
-          if (origTrack.parameters() == params &&
-              trk.trackWeight > m_cfg.minTrkWeight) {
-            // We expect that the i-th associated truth particle corresponds to
-            // the i-th track parameters
-            const auto& particle = associatedTruthParticles[i];
-            particleAtVtx.insert(particle);
-            SimBarcode vtxId =
-                particle.particleId().setParticle(0).setSubParticle(0);
-            contributingTruthVertices.push_back(vtxId);
-            foundMatchingParams = true;
+        // parameters. This allows us to identify the corresponding particle.
+        // TODO this should not be necessary if the tracks at vertex would keep
+        // this information
+        for (const auto& inputTrk : *tracks) {
+          const auto& params = inputTrk.parameters();
+
+          if (origTrack.parameters() == params) {
```

Comment:
This would simplify if we keep the tracks in the input tracks to the vertex.

Reply:
I refactored this whole matching stuff in another branch but it will take a few more PRs to get there